Fix guard for SUPPORTED_NODES

aorenste opened the pull request:
The special-case code for handling SUPPORTED_NODES was producing a guard that looked like:
```
"G['torch'].utils._pytree.SUPPORTED_NODES[<class '__main__.CausalLMOutputWithPast'>].type"
```
resulting in a eval error trying to evaluate the guard.

This change adds a new source type (`ClassSource`) which is given a class type (in this case `CausalLMOutputWithPast`) and attempts to fetch it from its defining module.  It then uses that to build the `SUPPORTED_NODES` guards instead of referring to the type directly.

Also added a unit test which fails before this change and passes after.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #120798



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

aorenste committed:
Fix guard for SUPPORTED_NODES

[ghstack-poisoned]

aorenste committed:
Update on "Fix guard for SUPPORTED_NODES"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 commented on torch/_dynamo/source.py:
```diff
@@ -322,9 +322,9 @@ def name(self):
         # 4) index is a constant - example string, integer
         if isinstance(self.index, Source):
-            if not isinstance(self.index, ConstDictKeySource):
+            if not isinstance(self.index, (ConstDictKeySource, ClassSource)):
```
I was hoping that we could just use the ConstDictKeySource to access the key. For a code like this (this passes on main)

~~~
import torch

class A:
    value = 5

d = {
    A : 2
}

@torch.compile(backend="eager")
def fn(x):
    return x * d[A]

fn(torch.randn(4))
~~~

This is the guard that we get ` ___check_obj_id(list(G['d'].keys())[0], 8593344)`

I am thinking can we do the same thing here with SUPPORTED_NODES. Basically when we do a get item on a ConstDictVariable (I am hoping SUPPORTED_NODES point to a ConstDictVariable), we can insert a ConstDictKeySource.

aorenste committed:
Update on "Fix guard for SUPPORTED_NODES"

The special-case code for handling SUPPORTED_NODES was producing a guard that looked like:
```
"G['torch'].utils._pytree.SUPPORTED_NODES[<class '__main__.CausalLMOutputWithPast'>].type"
```
resulting in a eval error trying to evaluate the guard.

This change adds a new source type (`ClassSource`) which is given a class type (in this case `CausalLMOutputWithPast`) and attempts to fetch it from its defining module.  It then uses that to build the `SUPPORTED_NODES` guards instead of referring to the type directly.

Also added a unit test which fails before this change and passes after.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

anijain2305 reviewed (approved):
Thanks for the fix! :) 
Looks much better.

anijain2305 commented on torch/_dynamo/variables/builder.py:
```diff
@@ -402,9 +402,11 @@ class Autotuner:
             result = {
                 ConstantVariable.create(k): UserDefinedObjectVariable(
                     v,
-                    source=GetItemSource(self.get_source(), k),
+                    source=GetItemSource(
+                        self.get_source(), ConstDictKeySource(self.get_source(), i)
+                    ),
                 )
-                for k, v in value.items()
+                for i, (k, v) in enumerate(value.items())
             }
```
This is pre-existing problem, but probably we should use LazyVariableTracker. Can you print TORCH_LOGS="guards" and check if we are guarding on all the keys of SUPPORTED_NODES even though we are only  using 1. If yes, then we should use LazyVariableTracker here. (I wont block this PR, this PR is definitely an improvement).

aorenste commented:
@pytorchbot merge

aorenste commented:
@pytorchbot merge -f "unrelated failures"

kit1980 commented:
@pytorchbot revert -m "the new test fails internally, see D54343456)" -c ghfirst

kit1980 commented:
@pytorchbot revert -m "the new test fails internally, see D54343456" -c ghfirst

aorenste committed:
Update on "Fix guard for SUPPORTED_NODES"

The special-case code for handling SUPPORTED_NODES was producing a guard that looked like:
```
"G['torch'].utils._pytree.SUPPORTED_NODES[<class '__main__.CausalLMOutputWithPast'>].type"
```
resulting in a eval error trying to evaluate the guard.

This change adds a new source type (`ClassSource`) which is given a class type (in this case `CausalLMOutputWithPast`) and attempts to fetch it from its defining module.  It then uses that to build the `SUPPORTED_NODES` guards instead of referring to the type directly.

Also added a unit test which fails before this change and passes after.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]
